File: src/cpp/cmai.cpp
#include <string>
#include <iostream>

// concept
template<class T>
concept bool Stringable = 
    requires(const T& a) {
        {to_string(a)} -> std::string;
    };

// modelling
class Person {
public:
    friend std::string to_string(const Person&) {return "hello";}
};

// algorithm
std::string bold(Stringable s) {
    return "<b>" + to_string(s) + "</b>";
}

Person make_person() {
    return Person{};
}

// instantiation
int main() {
    std::cout << bold(make_person());

    return 0;
}

File: src/haskell/cmai.hs
class Stringable a where
    toString :: a -> String

data Person = Person {firstname :: String, lastname :: String}

instance Stringable Person where
    toString p = take 1 (firstname p) ++ ". " ++ lastname p

bold :: (Stringable a) => a -> String
bold a = "<b>" ++ toString a ++ "</b>"

main = putStrLn $ bold $ Person "John" "Smith"
